### ISSUE TITLE
SSO auth renewal is blocking

### ISSUE THREAD

**Issue by cben** [openshift-online/ocm-sdk-go#232]
`tokensContext()` grabs a mutex:
https://github.com/openshift-online/ocm-sdk-go/blob/323cd4254e76dc7fde1ef703022bde9a61eeb002/token.go#L75-L79
IIUC this means that while a refresh/re-auth request to SSO is in progress, all concurrent API requests will be blocked.

This is sub-optimal — usually we renew some time before previous token expires, and could be used meanwhile without delaying requests.
Even the original request that triggered the renewal could proceed without blocking!

A second question is what happens with retries and concurrent requests.  The mutex is only held for a single try, but if a retry is needed it means we still don't have valid auth, so the 2nd request will immediately take the mutex; the first retry loop doesn't know that happened, so we can get several retry loops running in parallel. :twisted_rightwards_arrows: 
Once one of them succeeds, the following `tokensContext()` tries will all see valid auth and succeed immediately, and all retry loops will complete.

- [x] So this sounds _internally_ safe to me;
- [ ] however, externally we're not keeping the backoff we intended.  Are we risking a "thundering herd" problem where we make an SSO outage worse by amplifying the number of requests?
- [ ] also (very minor), this may confuse the retries logging/metrics #231.

I propose we move renewal off the requesting goroutine into a separate dedicated goroutine.  One per `Connection` object. This will simplify management of retries.
We need to decide whether renewal should wait for a request that needs auth (as done now, meaning zero traffic when unused), or be timer-based (minimizes latency).

@igoihman @nimrodshn @tzvatot @vkareh @zgalor @petli-openshift WDYT?

**Comment by petli-openshift** [comment]
overall sounds a good plan. I'ts probably better to still block the request if the access token is completely expired. it doesn't hurt to use a timer on each request. if it times out, it should fail the current request but not impact the renewal thread.
also for the renewal thread, is it possible to implement it in a way that it sends a single request to SSO no matter how many concurrent requests call the goroutine? the token could be delivered through a channel, so it decouples the renewal goroutine from the requests.

**Comment by tzvatot** [comment]
Access tokens live for 5 min. And it takes a few milliseconds to get it. So we are "paying" a few milliseconds every 5 minutes. I think that's a very small price to pay for the sake of thread safety. 
In other words - premature optimization IMO.

**Comment by cben** [comment]
Good point.  However, what if SSO is intermittently down, and we need retries?  That takes longer (backoff), and IIUC all this time no API requests will be made, despite auth being good.  IOW, we're not as "decoupled" from SSO as we hoped.

However, I'm fine with shelving this indefinitely, until actual need is proven :+1:.
For starters, we'll soon start collecting stats on retries and after few weeks we'll know something about actual impact.

**Comment by cben** [comment]
an easier change idea:
what if for some period before blocking for retries, we start "soft refresh" attempts, that do 1 request but don't retry, just fall back to using existing token?
(risk: This may also amplify SSO traffic without enforced backoff.)

hmm, actually that may already be happening!
The retries are around whole `tokensContext`, which tries a few strategies, but if those fail it warns and returns without error:
https://github.com/openshift-online/ocm-sdk-go/blob/b9d2a1d0dc0ae51b95188d7a8baa2e8307d82374/token.go#L152-L168
This should make `backoff.Retry` happy and not retry.
#### So blocking-for-retries scenario won't kick in until tokens are actually expired (at which point we must block anyway)?

@tzvatot is this last analysis correct?

**Comment by cben** [comment]
no, I don't think this happens now.  when `sendRefreshTokenForm()` or `sendRefreshTokenForm()` fail, it returns error immediately.
WDYT of changing that?

**Comment by cben** [comment]
token requests latency data from prod:
https://prometheus.app-sre-prod-04.devshift.net/graph?g0.range_input=1w&g0.expr=histogram_quantile(0.8%2C%20sum%20without(instance%2Cpod)%20(api_outbound_token_request_duration_bucket))&g0.tab=0&g1.range_input=1h&g1.expr=api_outbound_token_request_duration_bucket&g1.tab=1

The data is too coarse to take the quantile seriously, but what we know is this almost never completes ≤100ms :disappointed:  and almost always complete ≤1sec.
